executor/src/wasm_executor.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Rust implementation of Polkadot contracts.
+
+use parity_wasm::{deserialize_buffer, ModuleInstanceInterface, ProgramInstance};
+use parity_wasm::interpreter::{ItemIndex};
+use parity_wasm::RuntimeValue::{I32, I64};
+use std::collections::HashMap;
+use primitives::contract::CallData;
+use state_machine::{Externalities, CodeExecutor};
+use error::{Error, ErrorKind, Result};
+use std::sync::{Arc};
+use wasm_utils::{ModuleInstance, MemoryInstance, UserDefinedElements, AddModuleWithoutFullDependentInstance};
+
+struct Heap {
+	end: u32,
+}
+
+impl Heap {
+	fn new() -> Self {
+		Heap {
+			end: 1024,
+		}
+	}
+	fn allocate(&mut self, size: u32) -> u32 {
+		let r = self.end;
+		self.end += size;
+		r
+	}
+	fn deallocate(&mut self, _offset: u32) {
+	}
+}
+
+struct FunctionExecutor<'e, E: Externalities + 'e> {
+	heap: Heap,
+	memory: Arc<MemoryInstance>,
+	ext: &'e mut E,
+}
+
+impl<'e, E: Externalities> FunctionExecutor<'e, E> {
+	fn new(m: &Arc<ModuleInstance>, e: &'e mut E) -> Self {
+		FunctionExecutor {
+			heap: Heap::new(),
+			memory: Arc::clone(&m.memory(ItemIndex::Internal(0)).unwrap()),
+			ext: e,
+		}
+	}
+}
+
+impl_function_executor!(this: FunctionExecutor<'e, E>,
+	imported(n: u64) -> u64 => { println!("imported {:?}", n); n + 1 },
+	ext_memcpy(dest: *mut u8, src: *const u8, count: usize) -> *mut u8 => {
+		this.memory.copy_nonoverlapping(src as usize, dest as usize, count as usize).unwrap();
+		println!("memcpy {} from {}, {} bytes", dest, src, count);
+		dest
+	},
+	ext_memmove(dest: *mut u8, src: *const u8, count: usize) -> *mut u8 => { println!("memmove {} from {}, {} bytes", dest, src, count); dest },
+	ext_memset(dest: *mut u8, val: i32, count: usize) -> *mut u8 => { println!("memset {} with {}, {} bytes", dest, val, count); dest },
```

Comment:
this function and others are a little weird because although it returns pointers, those pointers are not actually safe to deref since they are just offsets in the heap memory instance. How is code meant to actually read from the written memory?

Reply:
so all function params are mapped to wasm-native types (`i8`/`u8`/`i32`/`i64`/`u32`/`u64`) for the code. most types are left alone, but pointers become `i32`s. the declared type is what is implemented as far as the rust compiler that generates the wasm code is concerned. it basically means you can copy/paste your rust-side extern declarations here and implement them.